llvm/lib/SYCLLowerIR/TargetHelpers.cpp

```diff
@@ -73,12 +66,16 @@ void KernelCache::populateKernels(Module &M) {
     return;
   }
 
-  // NVPTX kernels are identified by the global annotations metadata.
+  // NVPTX kernels are identified by their calling convention, and may have
+  // annotations.
   if (T.isNVPTX()) {
-    // Access `nvvm.annotations` to determine which functions are kernel
-    // entry points.
     auto *AnnotationMetadata = M.getNamedMetadata("nvvm.annotations");
-    // No kernels in the module, early exit.
+    for (auto &F : M) {
+      if (F.getCallingConv() == CallingConv::PTX_Kernel) {
+        Kernels.push_back(&F);
+	KernelData[&F] = KernelPayload{AnnotationMetadata};
+      }
+    }
     if (!AnnotationMetadata)
```

Comment:
I don't think this early-exit makes sense. If a module contains only `PTX_Kernel` functions and no `nvvm.annotations` metadata, we still want to collect all of those kernels and cache them.

Reply:
We do collect all the kernels and cache them (See line 75/76), the following code are for the DependentMD, no annotation means no dependent MD, so early exit here.

Reply:
Right you are, sorry about that. Perhaps we could add a comment explaining as much (that we're early-exiting as there are no dependent annotations)?

Reply:
Sure, done.